File: iOS/Pulsar/Sources/Pulsar/Lines/DiscreteLine.swift
import CoreHaptics


public struct DiscreteLine {
  private var events: [CHHapticEvent] = []
  
  public var getEvents: [CHHapticEvent] {
    return events
  }
  
  public mutating func addEvent(timestamp: Double, intensity: Float = 1, sharpness: Float = 1) {
    let event = CHHapticEvent(
      eventType: .hapticTransient,
      parameters: [
        CHHapticEventParameter(parameterID: .hapticIntensity, value: intensity),
        CHHapticEventParameter(parameterID: .hapticSharpness, value: sharpness)
      ],
      relativeTime: timestamp,
    )
    events.append(event)
  }
  
  public mutating func reset() {
    events.removeAll()
  }
}

File: kmp/Pulsar/library/src/iosMain/kotlin/com/swmansion/pulsar/Lines/DiscreteLine.kt
package com.swmansion.pulsar

import kotlinx.cinterop.ExperimentalForeignApi
import platform.CoreHaptics.CHHapticEvent
import platform.CoreHaptics.CHHapticEventParameter
import platform.CoreHaptics.CHHapticEventParameterIDHapticIntensity
import platform.CoreHaptics.CHHapticEventParameterIDHapticSharpness
import platform.CoreHaptics.CHHapticEventTypeHapticTransient

@OptIn(ExperimentalForeignApi::class)
internal class IOSDiscreteLine {
    private val events = mutableListOf<CHHapticEvent>()

    val getEvents: List<CHHapticEvent>
        get() = events

    fun addEvent(timestamp: Long, intensity: Float = 1f, sharpness: Float = 1f) {
        events += CHHapticEvent(
            eventType = CHHapticEventTypeHapticTransient,
            parameters = listOf(
                CHHapticEventParameter(CHHapticEventParameterIDHapticIntensity, intensity),
                CHHapticEventParameter(CHHapticEventParameterIDHapticSharpness, sharpness),
            ),
            relativeTime = timestamp.toDouble() / 1000.0,
        )
    }

    fun reset() {
        events.clear()
    }
}
